packages/core/admin/server/controllers/authentication.js

```diff
@@ -39,20 +41,36 @@ module.exports = {
         ctx.state.user = user;
 
         const sanitizedUser = getService('user').sanitizeUser(user);
-        strapi.eventHub.emit('admin.auth.success', { user: sanitizedUser, provider: 'local' });
+        // Multi factor authentication setting check
+        if (advanced.multi_factor_authentication) {
+          // Generate 6 digit code
+          const verificationCode = getService('token').createVerificationToken();
+          getService('auth').sendMultiFactorAuthenticationEmail({
+            user: sanitizedUser,
+            code: verificationCode
+          });
 
+          // Store the verification code and user information in the session for verification later
+          ctx.session.verificationCode = verificationCode;
+          ctx.session.user = user;
+        } else {
+          strapi.eventHub.emit('admin.auth.success', {user: sanitizedUser, provider: 'local'});
```

Comment:
I wanted to test this to see if my understanding was correct but I the MFA page doesn't load anymore after merging the main branch into my branch. Not sure what broke it..

But I think the reason why I didn't return auth success here is because you don't want the user to be authenticated before the 2FA code has also been verified. If I return the authentication, there is no reason for 2fa code. The user can modify the url to enter anywhere into the app after the initial login. At least that's how I remember implementing it couple of months ago :sweat_smile:

Reply:
I see I think it makes sense to make a new  emit  event for 2FA success and 2FA failed would be nice (My personal opinion as a community member). since now the end user can say hey you requested 2FA and it failed 3 times we mind need to lock the account in there custom code.

Reply:
@Boegie19 @markkaylor I was able run the app again today and ran some tests. I was actually wrong about the `strapi.eventHub.emit('admin.auth.success')` authenticating the user. I think the user is actually authenticated when I set the JWT token to the `ctx.body.data.token`. So I corrected that on my recent commits, to not create and pass the JWT token until the user has the 2FA code correct. 

@markkaylor Do you still want me to emit auth.success for both cases? And what are your thoughts on @Boegie19 's suggestion?

Reply:
I think boegie's idea sounds good 👍